src/thermo/Species.cpp

```diff
@@ -34,6 +37,46 @@ Species::~Species()
 {
 }
 
+double Species::molecularWeight() {
+    if (m_molecularWeight == Undef) {
+        double weight = 0.0;
+        const auto& elements = elementSymbolToWeight();
+        const auto& isotopes = isotopeSymbolToWeight();
+        for (const auto& comp : composition) {
+            auto search = elements.find(comp.first);
+            if (search != elements.end()) {
+                if (search->second < 0) {
+                    throw CanteraError("setMolecularWeight",
+                        "element '{}' has no stable isotopes", comp.first);
+                }
+                weight += search->second * comp.second;
+            } else {
+                search = isotopes.find(comp.first);
+                if (search != isotopes.end() && search->second > 0) {
+                    weight += search->second * comp.second;
+                }
+            }
+        }
+        setMolecularWeight(weight);
+    }
+    return m_molecularWeight;
+}
+
+void Species::setMolecularWeight(double weight) {
+    if (m_molecularWeight != Undef) {
+        double maxWeight = max(weight, m_molecularWeight);
+        double weight_cmp = fabs(weight - m_molecularWeight) / maxWeight;
+        if (weight_cmp > 1.0e-9) {
+            throw CanteraError(
+                "Species::setMolecularWeight",
+                "Molecular weight of a species cannot be changed."
+            );
```

Comment:
I believe this should raise a warning instead.

Reply:
Can you talk more about the advantages you see of a warning? I think allowing user code to overwrite the molecular weight would potentially be very confusing. The trade off that I see is people might find it confusing that sometimes the species they create can be added to a phase and sometimes they get this error.

Reply:
Another concern if this is a warning, what should the molecular weight be?

```
s = ct.Species(...)
s.molecular_weight
39.95
gas.add_species(s)  # with custom molecular weight
s.molecular_weight
????
```
Should it be the original value or the modified value? I can see cases where you'd want both behaviors...

Reply:
I agree that there is a tradeoff, where the reasoning you gave is ultimately more important from my perspective:

> The trade off that I see is people might find it confusing that sometimes the species they create can be added to a phase and sometimes they get this error.

Reply:
In your example, if the first `s.molecular_weight` is not called, then there is no error. Presumably, the user intent will be to use elements that are consistent with definitions of the phase the species is added to. I believe the most useful implementation here is to raise a warning and change the weight.

Reply:
I think raising an exception was a suggestion that came from me, but I'd agree that a warning if the weight changes would be sufficient for letting the user know that they're doing something that might lead to an unexpected result.